src/driver-steelseries.c

```diff
@@ -440,98 +508,84 @@ steelseries_write_led_v2(struct ratbag_led *led)
 {
 	struct ratbag_device *device = led->profile->device;
 	uint8_t buf[STEELSERIES_REPORT_SIZE] = {0};
-	uint16_t duration, min_duration;
 	int ret;
 
-	buf[0] = STEELSERIES_ID_LED;
-	buf[2] = led->index;
+	struct steelseries_cycle cycle;
+	struct steelseries_point point1;
+	struct steelseries_point point2;
+	struct steelseries_point point3;
+	struct steelseries_point point4;
 
-	/* not sure if these two are needed */
-	buf[15] = 0x01;
-	buf[17] = 0x01;
+	create_cycle(&cycle);
+	cycle.led_id = led->index;
 
 	switch(led->mode) {
 	case RATBAG_LED_OFF:
-		buf[19] = 0x01;
-		buf[27] = 0x01;
-		buf[28] = buf[31] = 0x00;
-		buf[29] = buf[32] = 0x00;
-		buf[30] = buf[33] = 0x00;
-		/* not sure why the duration is set for the steady color or why
-		   it is different for the two LEDs */
-		duration = led->index == 0 ? 10000 : 5000; /* 0x1027 or 0x8813 */
+		point1.color.red = 0x00;
+		point1.color.green = 0x00;
+		point1.color.blue = 0x00;
+		point1.pos = 0x00;
+
+		list_append(&cycle.points, &point1.link);
 		break;
 	case RATBAG_LED_ON:
-		buf[19] = 0x01;
-		buf[27] = 0x01;
-		buf[28] = buf[31] = led->color.red;
-		buf[29] = buf[32] = led->color.green;
-		buf[30] = buf[33] = led->color.blue;
-
-		/* not sure why the duration is set for the steady color or why
-		   it is different for the two LEDs */
-		duration = led->index == 0 ? 10000 : 5000; /* 0x1027 or 0x8813 */
+		point1.color = led->color;
+
+		list_append(&cycle.points, &point1.link);
 		break;
 	case RATBAG_LED_CYCLE:
-		buf[27] = 0x04; /* number of steps in cycle */
-
-		/* start color */
-		buf[28] = buf[31] = 0xFF;
-		buf[29] = buf[32] = 0x00;
-		buf[30] = buf[33] = 0x00;
-
-		/* Cycle to green */
-		buf[35] = 0x00;
-		buf[36] = 0xFF;
-		buf[37] = 0x00;
-		buf[38] = 0x54; /* normalized time share of animation */
-
-		/* Cycle to blue */
-		buf[39] = 0x00;
-		buf[40] = 0x00;
-		buf[41] = 0xFF;
-		buf[42] = 0x54; /* normalized time share of animation */
-
-		/* Cycle to red */
-		buf[43] = 0xFF;
-		buf[44] = 0x00;
-		buf[45] = 0x00;
-		buf[46] = 0x56; /* normalized time share of animation */
-
-		duration = led->ms;
+		point1.color.red = 0xFF;
+		point1.color.green = 0x00;
+		point1.color.blue = 0x00;
+		point1.pos = 0x00;
+
+		point2.color.red = 0x00;
+		point2.color.green = 0xFF;
+		point2.color.blue = 0x00;
+		point2.pos = 0x54;
+
+		point3.color.red = 0x00;
+		point3.color.green = 0x00;
+		point3.color.blue = 0xFF;
+		point3.pos = 0x54;
+
+		point4.color.red = 0xFF;
+		point4.color.green = 0x00;
+		point4.color.blue = 0x00;
+		point4.pos = 0x56;
```

Comment:
you're writing red twice? once at 0x00 and once at 0x56. Is this correct?

Reply:
Yes. When the cycle resets, it goes directly from 255 to 0. If the color isn't the same, it will just change, without any fade.